### ISSUE TITLE
Remove typescript dependency from core `@pulumi/pulumi` nodejs sdk

### ISSUE THREAD

**Issue by EvanBoyle** [pulumi/pulumi#7184]
Today our closure serialization libraries use typescript at runtime. That dependency alone is 50mb. We may want to consider isolating this dependency into a separate package outside of `@pulumi/pulumi` to make life easier for consumers (ie no crazy webpack configuration to shake this dependency out). This would involve moving the following into separate packages:

- dynamic providers
- sdk/node/runtime/closure

This would greatly contribute to improving https://github.com/pulumi/pulumi/issues/7182

**Comment by EvanBoyle** [comment]
Note that this is already causing us to implement strange behavior and workarounds in other areas like the pulumi github action: 

https://github.com/pulumi/actions/blob/master/scripts/replace-ts.js

**Comment by lukehoban** [comment]
Moving these would of course be a non-trivial breaking change.  Also, package like `@pulumi/aws` actually use closure serialization, so I expect they would need to import this new dependency anyway.  It might not be simple to unravel this.

An alternative would be to remove the `typescript` dependency entirely.  I am fairly certain it is only used to parse JavaScript code - and there are *much* lighter-weight options available for that.  It is used for:
* `parseFunction` - to identify lexically captured variables and whether a non-lexical this is referenced
* `rewriteSuperReferences` - to rewrite calls to `super(...)` though I don't recall exactly why this is even needed
* `isLegalFunctionName` - to determine if a string is a valid JS identifier

I believe none of these actually process TypeScript source code, and also that none of these use any type-directed logic.  So a much simpler JavaScript parser could likely be used as a replacement.

**Comment by EvanBoyle** [comment]
Perhaps we should consider making this change for 4.0. 

I dug up everything we import, many types but looks like a lot of functionality as well: 

```
import {
    ClassDeclaration, isClassDeclaration, ConstructorDeclaration, isConstructorDeclaration, SyntaxKind, isExpressionStatement, isArrowFunction, createSourceFile as tsCreateSourceFile, ScriptTarget,
    ScriptKind, SourceFile, Diagnostic, forEachChild, Node, CallExpression, FunctionLikeDeclarationBase, FunctionExpression, Identifier, MethodDeclaration, ThisExpression, Block, CatchClause, PropertyAssignment,
    PropertyAccessExpression, FunctionDeclaration, VariableDeclaration, ArrowFunction, ElementAccessExpression, isPropertyAccessExpression, isElementAccessExpression, isStringLiteral, StringLiteral, isCallExpression,
    isComputedPropertyName, BindingName, BindingPattern, isBindingElement, isVariableDeclarationList, BindingElement, NodeFlags, isIdentifier, isFunctionLike, NewLineKind, transform, createPrinter, EmitHint,
    TransformationContext, isFunctionDeclaration, visitEachChild, updateFunctionDeclaration, createIdentifier, createPropertyAccess, updatePropertyAccess, updateElementAccess, createThis, updateCall, visitNode, createScanner

} from "typescript";
```

**Comment by lukehoban** [comment]
> I dug up everything we import, many types but looks like a lot of functionality as well:

I am fairly certain the actual entry points that use any of this functionality are just the three I noted above - and that these are very narrow in what they do.  None of the types you list are referenced outside of these functions, so the usage is very encapsulated.

I think we should see if we can make a non-breaking change to replace the implementations of these three functions.  And if we cannot, then we could make a larger breaking change for 4.0 (or figure out a migration path of some sort ahead of 4.0?).

**Comment by EvanBoyle** [comment]
I see. Thanks for the clarification. 

> figure out a migration path of some sort ahead of 4.0

I could imagine accomplishing this by approaching it from the opposite side I originally suggested. We could create a `@pulumi/pulumi-core` package and move everything outside closure serialization into that package. and then re-export functionality from `@pulumi/pulumi-core` from `@pulumi/pulumi`. 

Tooling can then consume `@pulumi/pulumi-core`

**Comment by lukehoban** [comment]
> We could create a @pulumi/pulumi-core package and move everything outside closure serialization into that package. and then re-export functionality from @pulumi/pulumi-core from @pulumi/pulumi.

Yes - this is definitely another good option.  I'm still not sure how this would work for `@pulumi/aws` though - it itself references `serializeFunction` (which you are proposing moving/removing):  https://github.com/pulumi/pulumi-aws/blob/c27532148feb48ff5b9eaa49927b8c892c7f9831/sdk/nodejs/lambda/lambdaMixins.ts#L329:L329.  Would we have to create `@pulumi/aws-core` which is missing a handful of user-facing APIs (`CallbackFunction` and everything that optionally accepts a `CallbackFunction`)?

**Comment by komalali** [comment]
What if we only imported what was necessary from typescript instead of `import * as typescript from 'typescript'`? That might help with `ncc`.

**Comment by EvanBoyle** [comment]
@komalali I tried this and didn't see a substantial size reduction. This may be a result of the `ncc` tool though. Not sure how well tree shaking is supported/configured, and we may be able to get better results by using a custom webpack configuration.

**Comment by justinvp** [comment]
Doesn't `ts-node` require typescript as a peer dependency?

**Comment by dmattia** [comment]
Correct, ts-node requires `@types/node` and `typescript` as peer dependencies: https://github.com/TypeStrong/ts-node/blob/main/package.json#L141-L154

There is a PR that moves to make `typescript` a peer dependency though: https://github.com/pulumi/pulumi/pull/8199, which I believe would satisfy both sides here. It would strongly encourage installing some version of typescript, while not technically including it in the @pulumi/pulumi bundle.

**Comment by dmattia** [comment]
Any reason why https://github.com/pulumi/pulumi/pull/8199 was closed?
